### ISSUE TITLE
[BUG] - Autoscrolling does not work in some cases

### ISSUE THREAD

**Issue by serg-ty** [microsoft/playwright-java#727]
**Context:**
- Playwright Version: [playwright 17.1]
- Operating System: [Windows]
- Browser: [All]
- Extra: [any specific details about your environment]

**Code Snippet**

```java
/**
         * There are situations when several matching options available to select.
         * For example, 'Кардиология' и 'Детская кардиолоия' both match to locator
         * 'li >>text=Кардиология'.
         * Because of strict nature of locators there is need to select a strictly matching
         * option among available options.
         */
        public void clickStrictlyMatchingText(Locator input, String text) {
            input.click();
            input.fill(text);
            page.waitForSelector("li >> text=" + text,
                    new Page.WaitForSelectorOptions().setState(WaitForSelectorState.ATTACHED));
            Locator textToClick = page.locator("li >> text=" + text);
            int matchIndex = textToClick.allTextContents().indexOf(text);
            if (matchIndex != -1)
                textToClick.nth(matchIndex).click();
            else
                textToClick.click();
        }
```

**Describe the bug**

I need to select an item in edit box, which is formed as a request when I type some text into control. After I finish typing text a request is sent and matching options to select are attached to DOM as 'li' elements, and I need to click a needed option to finish my input. Control does not accept entered text unless some item ('li' element) is clicked explicitly.

I use default screen and viewport size, and autoscroll does not happen.

<img width="639" alt="Screenshot_366" src="https://user-images.githubusercontent.com/95364386/144396063-eace3b3d-d7ee-433d-ad26-bf3b12e13b7d.png">

Here is a screenshot taken after exception is thrown. It shows that matching options were available to select.

![screen](https://user-images.githubusercontent.com/95364386/144396447-69db2082-cf5e-424f-b9d8-0cce629996d0.jpg)

Here is some additional information from log:

<img width="262" alt="Screenshot_367" src="https://user-images.githubusercontent.com/95364386/144396675-e9b7cdd0-6857-4a43-a0e2-fba9dafc8353.png">

My problem is fixed if I provide different screen and viewport size

```java
context = browser.newContext(new Browser.NewContextOptions()
                        .setScreenSize(new ScreenSize(1900, 1000))
                        .setViewportSize(new ViewportSize(1900, 1000))
```

**Comment by yury-s** [comment]
This is likely a bug in Playwright but we need a repro. I assume you cannot share the page with us. Can you record a trace and share with us (you can send it over email if you don't want to make it public)?
```java
context.tracing().start(new Tracing.StartOptions().setSnapshots(true).setScreenshots(true));
// your code..
context.tracing().stop(new Tracing.StopOptions().setPath(Paths.get("/tmp/j.zip")));
```
If that's not an option can you tell us more about how the dropdown with the `<li>` is implemented?

**Comment by serg-ty** [comment]
Hi Yury,

I sent you a trace by e-mail.

**Comment by yury-s** [comment]
Thanks for sharing the trace, that's very helpful! So the problem is that the dropdown inside `<div class="react-tiny-popover-container">` is rendered as a fixed layout and `transform` which moves it outside of the viewport:

![image](https://user-images.githubusercontent.com/9798949/144661692-8b26d6a5-a196-49d1-a387-ea78dcaf6c78.png)

A human would not be able to click in the middle of such `<li>` element on the screen without scrolling it first. When the page is scrolled, the overlay is repositioned on the screen by some js code of the app that Playwright is not aware of (basically we cannot scroll an element with `position: fixed` by just sending mouse events to it). You can scroll the page manually, something like:
```java
page.evaluate("window.scrollBy(0, 100)");
```
Long term we can come up with some heuristics that would try scrolling the page when one is trying to click inside such elements with fixed position but it would all be just heuristics so for now for edge cases like this we suggest you manually scroll the page to ensure the popup renders inside the viewport.

**Comment by serg-ty** [comment]
Hi Yury,

Thanks for the feedback. I'm fine with manually scrolling a page, but I want to do this if it's required.
Is there a way to programmatically check the position of locator related to visible part of the screen?

I print some debug info about locator's bounding box height ```textToClick.boundingBox().height```

```
public void clickStrictlyMatchingText(Locator input, String text) {
            try {
                input.click();
                input.fill(text);
                page.waitForSelector("li >> text=" + text,
                        new Page.WaitForSelectorOptions()
                                .setState(WaitForSelectorState.ATTACHED)
                                .setTimeout(2000));
                clickMatchingTxt(text);
            } catch (TimeoutError e) {
                input.fill(text);
                page.waitForSelector("li >> text=" + text,
                        new Page.WaitForSelectorOptions()
                                .setState(WaitForSelectorState.ATTACHED));
                clickMatchingTxt(text);
            }
        }

        private void clickMatchingTxt(String text) {
            Locator textToClick = page.locator("li >> text=" + text);
            int matchIndex = textToClick.allTextContents().indexOf(text);
            if (matchIndex != -1) {
                System.out.println("selector position Y: " + textToClick.nth(matchIndex).boundingBox().height);
                System.out.println("page viewport position Y: " + page.viewportSize().height);
                textToClick.nth(matchIndex).click();
            } else {
                System.out.println("selector position Y: " + textToClick.boundingBox().height);
                System.out.println("page viewport position Y: " + page.viewportSize().height);
                textToClick.click();
            }
        }
```

But this appears to be not that I actually need.
<img width="201" alt="Screenshot_369" src="https://user-images.githubusercontent.com/95364386/145213047-4eaaee53-f085-49de-af93-a68a4f0e52bd.png">

My element to click is almost at the bottom of the page, but bounding box height is 29.

**Comment by ephung01** [comment]
Hello @serg-ty,
    Have you try send Press Enter Key after you entered your text in input?   This is not solving the playwright display "li", but could be work around to your issue.

**Comment by yury-s** [comment]
> Is there a way to programmatically check the position of locator related to visible part of the screen?

Yes, you already figured it out, [boundingBox](https://playwright.dev/java/docs/api/class-locator/#locator-bounding-box) is what you need.

> But this appears to be not that I actually need.

You are printing `height` while you need `y`:
```java
System.out.println("selector position Y: " + textToClick.nth(matchIndex).boundingBox().y);
```

**Comment by ephung01** [comment]
> Hello @serg-ty, Have you try send Press Enter Key after you entered your text in input? This is not solving the playwright display "li", but could be work around to your issue.

Alternatively... if you found "li" element you can try to use javascript click (ex: textToClick.evaluate("(e) => { e.click(); }"); which will not use coordinate mouse click (so it probably not affect by element inside or outside of viewport)

**Comment by serg-ty** [comment]
Hi @ephung01, thx for your suggestions and code examples, I will use them in my code if I run into some complex situations.

For now I applied ```page.evaluate("window.scrollBy(0, 50)");``` recommendation as it looks quite simple and works for my cases.

```
while (textToClick.boundingBox().y > page.viewportSize().height) {                       
    page.evaluate("window.scrollBy(0,50)");
}
```
